### DISCUSSION TITLE
`fatal error: 'linux/futex.h' file not found` when building jaxlib

### DISCUSSION THREAD

**Discussion by josephrocca** [pyodide/pyodide#2720] [Q&A]
This might be more suited to the emscripten discussion board, but I thought I'd ask here first in case this tends to come up with Pyodide stuff in particular.

I'm trying to build `jaxlib`, and it builds fine if I follow the normal (non-emscripten) instructions [here](https://jax.readthedocs.io/en/latest/developer.html), but when trying to use Pyodide's emscripten environment [as documented here](https://github.com/pyodide/pyodide/issues/2198#issuecomment-1153236401) I run into this error:
```
In file included from external/nsync/internal/common.c:18:
./external/nsync//platform/c++11.futex/platform.h:24:10: fatal error: 'linux/futex.h' file not found
#include <linux/futex.h>
         ^~~~~~~~~~~~~~~
```
Answers around the internet [imply](https://github.com/bitcoin/bitcoin/issues/16925#issuecomment-534372656) that I need to install some kernel stuff, but that's not the case because, as I said, it works fine if I build `jaxlib` normally (i.e. without emscripten). So this is something related to the emscripten environment that Pyodide is setting up. 

If you start the Docker container with `./run_docker --pre-built` and then create a single file called `test.c` with this content:
```
#include <linux/futex.h>
int main() { return 0; }
```
And then run this (I've just copied all the environment variables that Pyodide and Bazel add into a single `emcc` command to make it easier to replicate):
```
AR=ar     CARGO=cargo     CARGO_BUILD_TARGET=wasm32-unknown-emscripten     CARGO_HOME=/src/.docker_home/.cargo     CARGO_TARGET_WASM32_UNKNOWN_EMSCRIPTEN_LINKER=emcc     CC=cc     CPYTHONBUILD=/src/cpython/build/Python-3.10.2/     CXX=c++     EMSDK=/src/emsdk/emsdk     EMSDK_NODE=/src/emsdk/emsdk/node/14.18.2_64bit/bin/node     EM_CONFIG=/src/emsdk/emsdk/.emscripten     GCC=gcc     GFORTRAN=gfortran     HOME=/src/.docker_home     HOSTINSTALLDIR=/src/packages/.artifacts     HOSTSITEPACKAGES=/src/packages/.artifacts/lib/python3.10/site-packages     LC_CTYPE=C.UTF-8     LD=ld     NUMPY_LIB=/src/packages/.artifacts/lib/python3.10/site-packages/numpy/     PATH=/src/tools/symlinks:/src/emsdk/emsdk:/src/emsdk/emsdk/upstream/emscripten:/src/emsdk/emsdk/node/14.18.2_64bit/bin:/src/node_modules/.bin/:/src/.docker_home/.cargo/bin:/usr/local/bin:/usr/local/sbin:/usr/sbin:/usr/bin:/sbin:/bin:/src/packages/.artifacts/bin/     PKG_BUILD_DIR=/src/packages/jax/build/jax-0.3.13     PKG_CONFIG_PATH=/src/packages/.libs/lib/pkgconfig     PKG_VERSION=0.3.13     PWD=/proc/self/cwd     PYMAJOR=3     PYMICRO=2     PYMINOR=10     PYO3_CONFIG_FILE=/src/tools/pyo3_config.ini     PYODIDE=1     PYODIDE_JOBS=4     PYODIDE_PACKAGE_ABI=1     PYODIDE_ROOT=/src     PYTHONINCLUDE=/src/cpython/installs/python-3.10.2/include/python3.10     PYTHONPATH=/src/packages/.artifacts/lib/python3.10/site-packages     PYWASMCROSS_ARGS='{"pkgname": "jax", "cflags": "-msimd128\n -O2 -g0 -fPIC", "cxxflags": "-msimd128\n ", "ldflags": " -O2 -g0  -s MODULARIZE=1 -s LINKABLE=1 -s EXPORT_ALL=1 -s WASM=1 -std=c++14 -s LZ4=1 -L /src/cpython/installs/python-3.10.2/lib/ -s WASM_BIGINT  -s SIDE_MODULE=1", "target_install_dir": "/src/cpython/installs/python-3.10.2", "replace_libs": "", "builddir": "/src/packages/jax/build/jax-0.3.13", "PYTHONPATH": ["/src/packages/jax", "/usr/local/lib/python310.zip", "/usr/local/lib/python3.10", "/usr/local/lib/python3.10/lib-dynload", "/src/.docker_home/.local/lib/python3.10/site-packages", "/src/pyodide-build", "/src/.docker_home/.local/lib/python3.10/site-packages/tokenizers-0.12.1.dev0-py3.10-linux-x86_64.egg", "/usr/local/lib/python3.10/site-packages", "/src/cpython/installs/python-3.10.2/sysconfigdata"], "orig__name__": "pyodide_build.pywasmcross"}'     RUSTFLAGS='-C relocation-model=pic -C target-feature=+mutable-globals -C link-arg=-sSIDE_MODULE=1 -C link-arg=-sWASM_BIGINT'     SHLVL=0     SIDE_MODULE_CFLAGS='-O2 -g0 -fPIC'     SIDE_MODULE_LDFLAGS='-O2 -g0  -s MODULARIZE=1 -s LINKABLE=1 -s EXPORT_ALL=1 -s WASM=1 -std=c++14 -s LZ4=1 -L /src/cpython/installs/python-3.10.2/lib/ -s WASM_BIGINT  -s SIDE_MODULE=1'     STDLIB_MODULE_CFLAGS='-O2 -g0 -fPIC  -I Include/ -I .'     SYSCONFIG_NAME=_sysconfigdata__emscripten_wasm32-emscripten     TARGETINSTALLDIR=/src/cpython/installs/python-3.10.2     UNISOLATED_PACKAGES='["scipy", "pycparser", "cffi", "numpy"]'     WASM_LIBRARY_DIR=/src/packages/.libs     WASM_PKG_CONFIG_PATH=/src/packages/.libs/lib/pkgconfig     _=/usr/local/bin/python     _PYTHON_HOST_PLATFORM=emscripten_3_1_13_wasm32     _PYTHON_SYSCONFIGDATA_NAME=_sysconfigdata__emscripten_wasm32-emscripten /src/tools/symlinks/cc test.c
```
Then that replicates the issue. 

# Potential Solution:

As mentioned [here](https://github.com/pyodide/pyodide/issues/2198#issuecomment-1177184937), I think the way to solve this is to add `-I` flags to explicitly include directories with header files, like this: `emcc -I /usr/include/ -I /usr/include/x86_64-linux-gnu/ test.c`. I have no idea why it works fine without those when compiling with `cc`/`gcc`. Might be a bug with `emcc`?

**Comment by hoodmane** [comment]
Are you building with threading disabled? Either you haven't disabled threading correctly or this might be a bug in jax's support for threadless builds.

**Reply by josephrocca** [reply]
Ah I see, I didn't realise I had to do that - thanks! Looking through all the build files, I'm not sure if jax actually supports a non-threaded build. It does [seem to](https://github.com/google/jax/issues/743#issuecomment-536377029) support single-threaded mode at runtime. Also, [I was under the impression](https://emscripten.org/docs/porting/pthreads.html) that emscripten supports threaded stuff now that wasm supports threads? Or is that only in special circumstances?

By default `jax/build/build.py` builds a CPU version, and the high-level build options are [here](https://github.com/google/jax/blob/17de89b16ac5ee05aee03115d858e67489eab973/build/build.py#L314). The `--target_cpu_features` one initially sounds like the option I'm looking for, but it only supports 3 values:

> --target_cpu_features
> choices=["release", "native", "default"]
> help="What CPU features should we target? 'release' enables CPU features that should be enabled for a release build, which on x86-64 architectures enables AVX. 'native' enables -march=native, which generates code targeted to use all features of the current machine. 'default' means don't opt-in to any architectural features and use whatever the C compiler generates by default."

I'll learn about Bazel tomorrow and try to investigate JAX's [.bazelrc](https://github.com/google/jax/blob/main/.bazelrc) file. Worst case I can try to ask about this on the JAX discussion board and hopefully get a hint or two, or a message saying that it's not possible (hopefully that wouldn't be a dead end).

**Reply by hoodmane** [reply]
Wasm doesn't really support threads. Emscripten supports pthreads but it is a bit experimental and doesn't currently work with dynamic linking. Even in normal architectures, getting threads and dynamic linking to work together requires dark magic.

**Reply by hoodmane** [reply]
>  Worst case I can try to ask about this on the JAX discussion board

I would ask right away, often the maintainers/community are very helpful and can save you a lot of time.

**Reply by josephrocca** [reply]
[Okay, done](https://github.com/google/jax/discussions/11130). Thanks!

**Reply by josephrocca** [reply]
@hoodmane Okay so if I'm understanding this correctly, it sounds like it's not possible to build jaxlib without threading at this stage: https://github.com/google/jax/issues/11168#issuecomment-1162232086

Is this likely a dead end for now? I.e. until Emscripten threading doesn't require as much dark magic?

**Reply by ryanking13** [reply]
I'm not sure, but from my experience, the threading issue in Pyodide is mostly a runtime issue not a build time issue. So I think it is worth keep trying :). At least we can try "building" jax, though "running" it would throw an error regarding the threading issue. I don't have bandwidth to look into this currently, but I'll take a look when I have time.
